### PR TITLE
DRAFT: Experimentation with DKG - WT FROST

### PR DESCRIPTION
**Pull request by tbraun96**
The goal of this PR is to familiarize and to scope out how we might integrate the protocol with our existing DKG.

### CODE REVIEW

**Comment by drewstone** [comment]
This likely needs an update against the other PR which has moved files around?

File: dkg-gadget/src/wsts_proto/mod.rs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+		// Start by choosing signers. Since our indexes, in reality, will be based on the set of
+		// best authorities, we will choose the best of the best of authorities, so from 0..T
+		let mut signers = parties.lock().clone().into_iter().take(T as usize).collect::<Vec<_>>();
+
+		// Setup the signature aggregator
+		let mut sig_agg = wsts::v2::SignatureAggregator::new(K, T, public_key.clone()).unwrap();
+
+		// Each node will sign the message then broadcast their signature share to the other nodes
+		let (nonces, sig_shares, key_ids) =
```
[the remark this answers is not in the document]
Comment (drewstone) [reply dkg-gadget/src/wsts_proto/mod.rs:102]: If you're basing this purely off the repo's test utilities, it didn't seem like they had any networking involved anyway. It's purely to "run the protocol to completion". We wouldn't be using anything called `test_helpers` in production and so we would need to individually sign and gossip our signature.

It's worth noting that threshold Schnorr is different than ECDSA in that we just aggregate full signatures as is. Schnorr supports signature aggregation of signatures themselves just that we need to aggregate `t` of them.

[the remark this answers is not in the document]
Comment (xoloki) [reply dkg-gadget/src/wsts_proto/mod.rs:102]: This is correct.  The wsts repo is math/crypto only, no networking.  If you'd like to see an integration test of a fully networked implementation, you can see that here:

https://github.com/Trust-Machines/stacks-sbtc/blob/main/frost-coordinator/src/coordinator.rs#L551

This spawns a `frost-coordinator` and a configurable number of `frost-signer`s which do a DKG round followed by a signing round.  There are network packets defined here:

https://github.com/Trust-Machines/stacks-sbtc/blob/main/frost-signer/src/signing_round.rs#L149

These packets are all signed with ECDSA, and the DKG private shares are encrypted for the rcpt using DH shared keys derived from the signers' ECDSA keys.

As noted in the comment above, the signature shares can be aggregated trivially.  However, the signers need to know what the signing set is before they can construct their signature shares (due to the way Lagrange interpolation works)

[the remark this answers is not in the document]
Comment (drewstone) [reply dkg-gadget/src/wsts_proto/mod.rs:102]: > However, the signers need to know what the signing set is before they can construct their signature shares (due to the way Lagrange interpolation works)

@xoloki this seems similar to GG protocols. Is there no protocol that relaxes this? Where t out of any of the N signers sign and clients can choose to aggregate arbitrarily the first t they find? Just curious, as you've stated is how we've built our infra so should still work out.

[the remark this answers is not in the document]
Comment (xoloki) [reply dkg-gadget/src/wsts_proto/mod.rs:102]: > > However, the signers need to know what the signing set is before they can construct their signature shares (due to the way Lagrange interpolation works)
> 
> @xoloki this seems similar to GG protocols. Is there no protocol that relaxes this? Where t out of any of the N signers sign and clients can choose to aggregate arbitrarily the first t they find? Just curious, as you've stated is how we've built our infra so should still work out.

As I discussed in the WSTS paper, there is one robust protocol which allows completion as long as a threshold completes signing:  https://dl.acm.org/doi/10.5555/646038.678297

The price you pay for robustness is that every signature requires running a full DKG to produce a shared nonce, and multiple rounds of communication.  This is okay if you have a small number of signers, but if you have hundreds or thousands it quickly becomes untenable.  For WSTS we chose to optimize vanilla FROST instead, and use a meta-protocol on top to achieve robustness.  For cryptocurrency applications you can detect and sanction any malicious actors, so it makes sense to optimize for the case where there are none.

**Pull request closed without merging**